### PR TITLE
Added reconnection abilities

### PR DESCRIPTION
**Pull request by Faibk**
This pull request implements the behaviour requested in issue #246 and enables roslibjs to be used in more dynamic scenarios, e.g. single-page applications, or under unstable network conditions. If the websocket connection breaks down all current topic subscriptions and advertisements are restored on reconnection. 

To produce a stable and reliable result without memory leaks it is necessary to always unsubscribe topics  not longer needed. Before this PR this was possible for topics only and not as critical since the connection needed to be reinstatiated and the garbage collector removed all handlers anyway. Reconnection reuses the ros connection object, so this pull request adds the ability to dispose `ActionClients` and `TFClients` (unsubscribe and unadvertise all internal topics). Also to prevent multiple resubscriptions on failed reconnection attempts the `callOnceConnection` method has been introduced which only queues up a particular message once.

Also the `Goal` class has been modifed to accept setting the goalID manually (this is required for use cases in which goals need to be restored from the action server on reconnection).

### Compatibility Notes
This PR does only add handlers for the case of reconnection, it does not contain the reconnection mechanism itself. That means this should not break any current use of the library following the usual reinstantiation/window reload approach. Only intentionally calling `.connect()` on the instantiated ros-websocket again will trigger the reconnection/resubscription handlers introduced in this PR.

### CODE REVIEW

File: src/actionlib/ActionClient.js [hunk 1]
```diff
@@ -129,4 +129,15 @@ ActionClient.prototype.cancel = function() {
   this.cancelTopic.publish(cancelMessage);
 };
 
+/**
+ * Unsubscribe and unadvertise all topics associated with this ActionClient.
+ */
+ActionClient.prototype.dispose = function() {
```
Comment (jihoonl) [inline src/actionlib/ActionClient.js]: This looks good to me thanks.

File: src/core/Ros.js [hunk 2]
```diff
@@ -135,6 +136,33 @@ Ros.prototype.callOnConnection = function(message) {
 };
 
 /**
+ * Sends the message over the WebSocket, but queues the message up if not yet
+ * connected. Only queues a particular message once.
+ */
+Ros.prototype.callOnceConnection = function(message) {
```
[the remark this answers is not in the document]
Comment (jihoonl) [reply src/core/Ros.js]: I have implemented reconnection behavior in different way in [jihoonl/reconnect](https://github.com/jihoonl/roslibjs/commit/082d6c6d6e0681eec734d814c2c4396d83afadb4). Could you take a look if this works for you? As the comment above, I think this behavior should be handled within Topic.js level.

**Comment by Faibk** [comment]
I made the requested changes, you are right reconnection is indeed only a matter of topic.js
I pulled your code and my app runs fine on it. I also did not find any loopholes, so I think this is good now.

Thank you for the extensive feedback, this was my first major pull request :)

**Comment by jihoonl** [comment]
always appreciates PRs :wink: :wink:. It would be nice if you can create a PR per a feature. e.g) one PR to add `dispose()`. another PR to add reconnect capability.

**Pull request merged**